File: app/src/main/java/com/md/modesetters/LearningModeSetter.java
package com.md.modesetters;

import android.app.Activity;
import android.app.AlertDialog;
import android.content.DialogInterface;
import android.os.SystemClock;
import android.view.KeyEvent;
import android.view.View;
import android.view.ViewGroup;
import android.widget.Button;
import android.widget.TextView;

import com.md.AudioPlayer;
import com.md.DbNoteEditor;
import com.md.DbRepEditor;
import com.md.ModeHandler;
import com.md.R;
import com.md.RevisionQueue;
import com.md.SpacedRepeaterActivity;
import com.md.provider.AbstractRep;
import com.md.provider.Note;
import com.md.utils.ScreenDimmer;
import com.md.utils.ToastSingleton;

import org.jetbrains.annotations.Nullable;

import java.util.ArrayDeque;
import java.util.ArrayList;
import java.util.Deque;
import java.util.List;
import java.util.concurrent.TimeUnit;

import static com.md.SpacedRepeaterActivity.PRESS_GROUP_MAX_GAP_MS_INSTANT;

public class LearningModeSetter extends ModeSetter implements
        ItemDeletedHandler {

    private static LearningModeSetter instance = null;

    protected LearningModeSetter() {
    }

    public static LearningModeSetter getInstance() {
        if (instance == null) {
            instance = new LearningModeSetter();
        }
        return instance;
    }

    /**
     * @param memoryDroid
     * @param modeHand
     */
    public void setUp(SpacedRepeaterActivity memoryDroid, ModeHandler modeHand) {
        parentSetup(memoryDroid, modeHand);
        this.memoryDroid = memoryDroid;
    }

    private final Deque<Note> lastNoteList = new ArrayDeque<>();

    private AbstractRep lastNoteRep;
    private SpacedRepeaterActivity memoryDroid;
    private Note currentNote;
    private int originalSize;
    private int repCounter;
    private int missCounter;
    private boolean questionMode = true;

    public void setupModeImpl(final Activity context) {
        originalSize = RevisionQueue.getCurrentDeckReviewQueue().getSize();

        lastNoteList.clear();

        commonSetup(context, R.layout.learnquestion);

        // Let's just load up the learn question also to get it ready.
        setupQuestionMode(context);
    }

    protected void updateShouldRepeat() {
        AudioPlayer.getInstance().setShouldRepeat(true);
    }

    private void commonLayoutSetup() {
        ViewGroup gestures = this.memoryDroid
                .findViewById(R.id.gestures);
        gestures.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View view) {
                mActivity.handleRhythmUiTaps(LearningModeSetter.this, SystemClock.uptimeMillis(), SpacedRepeaterActivity.PRESS_GROUP_MAX_GAP_MS_SCREEN);
            }
        });

        final TextView reviewNumber = (TextView) this.memoryDroid
                .findViewById(R.id.reviewNumber);

        String firstLine = "Scheduled: " + originalSize;
        firstLine += "\nPerformed: " + repCounter;
        String secondLine = "\nItems Missed: " + missCounter;
        secondLine += "\nRemaining: " + RevisionQueue.getCurrentDeckReviewQueue().getSize();
        if (currentNote != null) {
            secondLine += "\n\nEasiness: " + currentNote.getEasiness();
            secondLine += "\nInterval: " + currentNote.interval();
        }

        reviewNumber.setText("::::" + ((questionMode) ? "Question" : "Answer") + "::::\n"
                + firstLine + "\n" + secondLine);

        final DbNoteEditor noteEditor = DbNoteEditor.getInstance();

        Button deleteButton = memoryDroid
                .findViewById(R.id.deleteButton);

        deleteButton.setOnClickListener(new DeleterOnClickListener(noteEditor, mActivity, this));

        memoryDroid.findViewById(R.id.new_note_button).setOnClickListener(new MultiClickListener() {
            @Override
            public void onMultiClick(View v) {
                CreateModeSetter.getInstance().setupMode(memoryDroid);
            }
        });

        memoryDroid.findViewById(R.id.dim_screen_button).setOnClickListener(new MultiClickListener() {
            @Override
            public void onMultiClick(View v) {
                toggleDim();
            }
        });

        memoryDroid.findViewById(R.id.back_button).setOnClickListener(new MultiClickListener() {
            @Override
            public void onMultiClick(View v) {
                mActivity.onBackPressed();
            }
        });

        View audioFocusToggle = memoryDroid.findViewById(R.id.audio_focus_toggle);
        audioFocusToggle.setOnKeyListener(
                new View.OnKeyListener() {
                    @Override
                    public boolean onKey(View view, int keyCode, KeyEvent keyEvent) {
                        // BR301 sends an enter command, which we want to ignore.
                        if (keyCode == KeyEvent.KEYCODE_ENTER) {
                            return true;
                        }
                        return false;
                    }
                }
        );

        audioFocusToggle.setOnClickListener(new MultiClickListener() {
            @Override
            public void onMultiClick(View v) {
                // Release audio focus since the dialog prevents keyboards from controlling memprime.
                mActivity.maybeChangeAudioFocus(!mActivity.hasAudioFocus());
            }
        });
    }

    public void handleTapCount(int count) {
        memoryDroid.handleRhythmUiTaps(this, SystemClock.uptimeMillis(), PRESS_GROUP_MAX_GAP_MS_INSTANT, count);
    }

    /**
     * Note: currently this is hardcoded to 3 clicks.
     */
    public static abstract class MultiClickListener implements View.OnClickListener {

        private long mClickWindowStartMillis = 0L;
        private int mClickCount = 0;

        public abstract void onMultiClick(View v);

        @Override
        public void onClick(View v) {
            final long currentTimeMillis = SystemClock.uptimeMillis();
            if (currentTimeMillis - mClickWindowStartMillis < TimeUnit.SECONDS.toMillis(1)) {
                mClickCount++;
                if (mClickCount == 3) {
                    onMultiClick(v);
                }
            } else {
                // Start new window.
                mClickWindowStartMillis = currentTimeMillis;
                mClickCount = 1;
            }
        }
    }

    public void undo() {
        if (questionMode) {
            undoLastQuestion(mActivity);
        } else {
            undoThisQuestion(mActivity);
        }
    }

    public void handleReplay() {
        mActivity.keepHeadphoneAlive();
        if (questionMode) {
            replay();
        } else {
            replayA();
        }
        hideSystemUi();
    }

    public void proceedFailure() {
        if (questionMode) {
            proceed(mActivity);
        } else {
            updateScoreAndMoveToNext(mActivity, 1);
        }
        mActivity.keepHeadphoneAlive();
    }

    public String secondaryAction() {
        if (questionMode) {
            handleReplay();
            return "bad go";
        } else {
            proceedFailure();
            return "bad bad";
        }
    }

    public void proceed() {
        ScreenDimmer.getInstance().keepScreenOn(mActivity);
        if (questionMode) {
            proceed(mActivity);
        } else {
            updateScoreAndMoveToNext(mActivity, 4);
        }
        hideSystemUi();
        mActivity.keepHeadphoneAlive();
    }

    private void setupQuestionMode(final Activity context,
                                   boolean shouldUpdateQuestion) {
        if (shouldUpdateQuestion) {
            updateVal();
        }
        memoryDroid.setContentView(R.layout.learnquestion);
        applyDim(mIsDimmed);
        questionMode = true;

        if (currentNote != null) {
            AudioPlayer.getInstance().playFile(currentNote.getQuestion(), null, true);
        } else {
            if (mActivity instanceof SpacedRepeaterActivity) {
                // Release audio focus since the dialog prevents keyboards from controlling memprime.
                mActivity.maybeChangeAudioFocus(false);
            }
            AlertDialog.Builder builder = new AlertDialog.Builder(context);

            TtsSpeaker.speak("Great job! Deck done.");
            builder.setMessage("Great job! Deck done.")
                    .setCancelable(false)
                    .setNeutralButton("Ok",
                            new DialogInterface.OnClickListener() {
                                public void onClick(DialogInterface dialog,
                                                    int id) {

                                }
                            });
            AlertDialog alert = builder.create();
            alert.show();

        }

        commonLayoutSetup();

        memoryDroid.findViewById(R.id.rerecord)
                .setOnTouchListener(
                        new RecordOnClickListener(currentNote, context, false,
                                getLastOrNull()));

    }

    @Nullable
    private Note getLastOrNull() {
        return !lastNoteList.isEmpty() ? lastNoteList.getLast() : null;
    }

    private void setupQuestionMode(final Activity context) {
        setupQuestionMode(context, true);
    }

    private void setupAnswerMode(final Activity context) {

        questionMode = false;
        memoryDroid.setContentView(R.layout.learnquestion);
        applyDim(mIsDimmed);

        if (currentNote != null) {
            AudioPlayer.getInstance().playFile(currentNote.getAnswer(), null, true);
        }

        commonLayoutSetup();

        memoryDroid.findViewById(R.id.rerecord)
                .setOnTouchListener(
                        new RecordOnClickListener(currentNote, context, true,
                                getLastOrNull()));
    }

    private void updateVal() {
        currentNote = RevisionQueue.getCurrentDeckReviewQueue().peekQueue();
        if (currentNote != null) {
            repCounter++;
        }
    }

    private void updateScoreAndMoveToNext(Activity context, int newGrade) {
        if (currentNote != null) {
            applyGradeStatic(context, newGrade, currentNote);
            ToastSingleton.getInstance().msg("Easiness: " + currentNote.getEasiness() + " Interval " + currentNote.getInterval());
            setupQuestionMode(context);
        }
    }

    public void applyGradeStatic(Activity context, int newGrade,
                                 Note currentNote) {
        DbNoteEditor noteEditor = DbNoteEditor.getInstance();

        if (lastNoteRep != null && getLastOrNull() != null) {
            DbRepEditor repEditor = DbRepEditor.getInstance();
            repEditor.insert(lastNoteRep);
        }

        lastNoteList.addLast(currentNote.clone());
        // Create the rep info before updating the note with the new internal.
        lastNoteRep = new AbstractRep(currentNote.getId(),
                currentNote.getInterval(), newGrade, System.currentTimeMillis());

        currentNote.process_answer(newGrade);

        noteEditor.update(context, currentNote);

        // If you scored too low review it again, at the end.
        if (currentNote.is_due_for_acquisition_rep()) {
            RevisionQueue.getCurrentDeckReviewQueue().updateNote(currentNote, false);
            missCounter++;
        } else {
            RevisionQueue.getCurrentDeckReviewQueue().removeNote(currentNote.getId());
        }
    }

    @Override
    public void deleteNote() {
        final DbNoteEditor noteEditor = DbNoteEditor.getInstance();
        Note note = currentNote;
        if (note != null) {
            noteEditor.deleteCurrent(mActivity, note);
            RevisionQueue.getCurrentDeckReviewQueue().removeNote(note.getId());
        }

        setupQuestionMode(mActivity);
    }

    private void replayA() {
        ScreenDimmer.getInstance().keepScreenOn(mActivity);
        if (currentNote != null) {
            AudioPlayer.getInstance().playFile(currentNote.getAnswer(), null, true);
        }
    }

    private void replay() {
        ScreenDimmer.getInstance().keepScreenOn(mActivity);
        if (currentNote != null) {
            AudioPlayer.getInstance().playFile(currentNote.getQuestion(), null, true);
        }
    }

    private void proceed(final Activity context) {
        if (currentNote != null) {
            setupAnswerMode(context);
        }
    }

    private void undoLastQuestion(final Activity context) {
        if (!lastNoteList.isEmpty()) {
            currentNote = lastNoteList.removeLast();
            repCounter--;
            DbNoteEditor noteEditor = DbNoteEditor.getInstance();
            noteEditor.update(context, currentNote);
            // In case the grade was bad take it out of revision queue.
            RevisionQueue.getCurrentDeckReviewQueue().removeNote(currentNote.getId());
            RevisionQueue.getCurrentDeckReviewQueue().addToFront(currentNote);
            setupAnswerMode(context);
        } else {
            TtsSpeaker.speak("Nothing to undo");
        }
    }

    public void resetActivity() {
        setupMode(mActivity);
    }

    private void undoThisQuestion(final Activity context) {
        setupQuestionMode(context, false);
    }

    protected void adjustScreenLock() {
        ScreenDimmer.getInstance().keepScreenOn(mActivity);
        hideSystemUi();
        mIsDimmed = false;
    }

    private boolean mIsDimmed = false;

    public void toggleDim() {
        mIsDimmed = !mIsDimmed;
        applyDim(mIsDimmed);
        showSystemUi();
    }

    public void mark() {
        if (currentNote != null) {
            currentNote.setMarked(true);
            final DbNoteEditor noteEditor = DbNoteEditor.getInstance();
            noteEditor.update(mActivity, currentNote);
            ToastSingleton.getInstance().msg("Marked note");
        }
    }

    private void applyDim(boolean isDimmed) {
        // Note making the root invisible makes the background grey.
        ViewGroup layout = this.memoryDroid.findViewById(R.id.learn_layout_root);
        for (int i = 0; i < layout.getChildCount(); i++) {
            applyDimToView(isDimmed, layout.getChildAt(i));
        }

        applyDimToView(isDimmed, this.memoryDroid.findViewById(R.id.reviewNumber));

        // Gestures listeners don't work if dimmed.
        applyDimToView(false, this.memoryDroid.findViewById(R.id.gestures));
    }

    private void applyDimToView(boolean isDimmed, View view) {
        if (!isDimmed) {
            view.setVisibility(View.VISIBLE);
        } else {
            view.setVisibility(View.INVISIBLE);
        }
    }
}

File: app/src/main/java/com/md/modesetters/LearningModeSetter.kt
package com.md.modesetters

import android.app.Activity
import android.os.SystemClock
import android.provider.MediaStore.Audio
import android.view.View
import android.view.ViewGroup
import android.widget.Button
import android.widget.TextView
import androidx.lifecycle.lifecycleScope
import com.md.*
import com.md.RevisionQueue.Companion.currentDeckReviewQueue
import com.md.provider.AbstractRep
import com.md.provider.Note
import com.md.utils.ScreenDimmer
import com.md.utils.ToastSingleton
import com.md.workers.BackupPreferences.markAllStale
import dagger.hilt.android.scopes.ActivityScoped
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.delay
import kotlinx.coroutines.isActive
import kotlinx.coroutines.launch
import java.util.*
import java.util.concurrent.TimeUnit
import javax.inject.Inject

@ActivityScoped
class LearningModeSetter @Inject constructor() : ModeSetter(), ItemDeletedHandler {
    /**
     * @param memoryDroid
     * @param modeHand
     */
    fun setUp(memoryDroid: SpacedRepeaterActivity?, modeHand: ModeHandler?) {
        parentSetup(memoryDroid, modeHand)
        this.memoryDroid = memoryDroid
    }

    private val lastNoteList: Deque<Note> = ArrayDeque()
    private var lastNoteRep: AbstractRep? = null
    private var memoryDroid: SpacedRepeaterActivity? = null
    private var currentNote: Note? = null
    private var originalSize = 0
    var repCounter = 0
    private var missCounter = 0
    private var questionMode = true
    override fun switchModeImpl(context: Activity) {
        originalSize = currentDeckReviewQueue!!.getSize()
        lastNoteList.clear()
        commonSetup(context, R.layout.learnquestion)

        // Let's just load up the learn question also to get it ready.
        setupQuestionMode(context)
    }

    var middleTappableButton: View? = null

    private fun commonLayoutSetup() {
        val memoryDroid = memoryDroid!!
        middleTappableButton = memoryDroid
            .findViewById<ViewGroup>(R.id.gestures)
        middleTappableButton?.setOnClickListener { _ ->
            mActivity!!.handleRhythmUiTaps(
                this@LearningModeSetter,
                SystemClock.uptimeMillis(),
                SpacedRepeaterActivity.PRESS_GROUP_MAX_GAP_MS_SCREEN
            )
        }
        middleTappableButton?.setOnLongClickListener { _ ->
            if (AudioPlayer.instance.wantsToPlay) {
                AudioPlayer.instance.pause()
            } else {
                AudioPlayer.instance.playWhenReady()
            }

            true
        }
        val reviewNumber = memoryDroid
            .findViewById<View>(R.id.reviewNumber) as TextView
        var firstLine = "Scheduled: $originalSize"
        firstLine += "\nPerformed: $repCounter"
        var secondLine = "\nItems Missed: $missCounter"
        secondLine += """
            
            Remaining: ${currentDeckReviewQueue!!.getSize()}
            """.trimIndent()
        if (currentNote != null) {
            secondLine += """
                
                
                Easiness: ${currentNote!!.easiness}
                """.trimIndent()
            secondLine += """
                
                Interval: ${currentNote!!.interval()}
                """.trimIndent()
        }
        reviewNumber.text = """
            ::::${if (questionMode) "Question" else "Answer"}::::
            $firstLine
            $secondLine
            """.trimIndent()
        val noteEditor = DbNoteEditor.instance
        val deleteButton = memoryDroid
            .findViewById<Button>(R.id.deleteButton)
        deleteButton.setOnClickListener(DeleterOnClickListener(noteEditor, mActivity, this))
        memoryDroid.findViewById<View>(R.id.new_note_button)
            .setOnClickListener(object : MultiClickListener() {
                override fun onMultiClick(v: View?) {
                    CreateModeSetter.switchMode(memoryDroid!!)
                }
            })
        memoryDroid.findViewById<View>(R.id.dim_screen_button)
            .setOnClickListener(object : MultiClickListener() {
                override fun onMultiClick(v: View?) {
                    toggleDim()
                }
            })
        memoryDroid.findViewById<View>(R.id.back_button)
            .setOnClickListener(object : MultiClickListener() {
                override fun onMultiClick(v: View?) {
                    mActivity!!.onBackPressed()
                }
            })
        val audioFocusToggle = memoryDroid!!.findViewById<View>(R.id.audio_focus_toggle)
        audioFocusToggle.setOnClickListener { v: View? -> AudioPlayer.instance.playWhenReady() }
    }

    fun handleTapCount(count: Int) {
        memoryDroid!!.handleRhythmUiTaps(
            this,
            SystemClock.uptimeMillis(),
            SpacedRepeaterActivity.PRESS_GROUP_MAX_GAP_MS_INSTANT,
            count
        )
    }

    /**
     * Note: currently this is hardcoded to 3 clicks.
     */
    abstract class MultiClickListener : View.OnClickListener {
        private var mClickWindowStartMillis = 0L
        private var mClickCount = 0
        abstract fun onMultiClick(v: View?)
        override fun onClick(v: View) {
            val currentTimeMillis = SystemClock.uptimeMillis()
            if (currentTimeMillis - mClickWindowStartMillis < TimeUnit.SECONDS.toMillis(1)) {
                mClickCount++
                if (mClickCount == 3) {
                    onMultiClick(v)
                }
            } else {
                // Start new window.
                mClickWindowStartMillis = currentTimeMillis
                mClickCount = 1
            }
        }
    }

    override fun undo() {
        if (questionMode) {
            if (!AudioPlayer.instance.wantsToPlay) {
                AudioPlayer.instance.playWhenReady()
            } else {
                undoLastQuestion(mActivity!!)
            }
        } else {
            undoThisQuestion(mActivity!!)
        }
    }

    override fun handleReplay() {
        if (questionMode) {
            replay()
        } else {
            replayA()
        }
        hideSystemUi()
    }

    override fun proceedFailure() {
        if (questionMode) {
            proceed(mActivity!!)
        } else {
            updateScoreAndMoveToNext(mActivity!!, 1)
        }
    }

    override fun secondaryAction(): String {
        val messageToSpeak = if (questionMode) {
            if (AudioPlayer.instance.hasPlayedCurrentFile()) {
                return ""
            }

            "secondary action. proceed"
        } else {
            "bad bad"
        }
        proceedFailure()
        return messageToSpeak
    }

    /** Moves this note to the end of the queue.  */
    override fun postponeNote(shouldQueue: Boolean) {
        val currentNote = currentNote ?: return

        if (shouldQueue) {
            // Place at end of queue.
            currentDeckReviewQueue!!.updateNote(currentNote, false)
        } else {
            val editor = DbNoteEditor.instance
            val context = mActivity
            if (context != null && editor != null) {
                currentNote.decreasePriority()
                editor.update(context, currentNote)
            } else {
                TtsSpeaker.speak("Error decreasing priority")
            }

            currentDeckReviewQueue!!.hardPostpone(currentNote)
        }
        // Prepare the next note in the queue.
        setupQuestionMode(mActivity!!)
    }

    override fun proceed() {
        ScreenDimmer.getInstance().keepScreenOn(mActivity)
        if (!AudioPlayer.instance.hasPlayedCurrentFile()) {
          return
        }
        if (questionMode) {
            proceed(mActivity!!)
        } else {
            updateScoreAndMoveToNext(mActivity!!, 4)
        }
        hideSystemUi()
    }

    private fun setupQuestionMode(
        context: Activity,
        shouldUpdateQuestion: Boolean = true
    ) {
        MoveManager.recordQuestionProceed()
        if (shouldUpdateQuestion) {
            updateVal()
        }
        memoryDroid!!.setContentView(R.layout.learnquestion)
        applyDim(mIsDimmed)
        questionMode = true

        val lifeCycleOwner = mActivity
        val currentNote = currentNote
        if (currentNote != null && lifeCycleOwner != null) {
            val questionAudioFilePath = currentNote.question
            var shouldPlayTwiceInARow = true
            MoveManager.replaceMoveJobWith(lifeCycleOwner.lifecycleScope.launch(Dispatchers.Main) {
                while (isActive) {
                    if (questionAudioFilePath != currentNote.question) {
                        // TODOJ maybe turn into precondition.
                        return@launch
                    }
                    while (middleTappableButton?.isPressed == true) {
                        delay(500)
                        continue
                    }
                    if (!mActivity.isAtLeastResumed()) {
                        return@launch
                    }

                    AudioPlayer.instance.playFile(
                        questionAudioFilePath,
                        firedOnceCompletionListener = {},
                        shouldRepeat = shouldPlayTwiceInARow,
                        autoPlay = true)

                    // Delay for 1 second to avoid interfering with the main file that is playing...
                    delay(1_000)
                    // ... then preload a file that is likely to be used soon.
                    try {
                        currentDeckReviewQueue!!.preload()
                    } catch (e: IllegalStateException) {
                        TtsSpeaker.speak("preload failed", lowVolume = true)
                        e.printStackTrace()
                    }

                    // TODO(jrochest) Start the delay below after done playing. Utilize the
                    // firedOnceCompletionListener
                    // Note this works fine if it replays before the note is done.
                    shouldPlayTwiceInARow = false
                    delay(30_000)
                    // This TTS is mostly helpful to avoid the bluetooth speakers being off during
                    // replay.
                    TtsSpeaker.speak("replay", lowVolume = true)
                    delay(500)
                }
            })
        } else {
            if (mActivity != null) {
                // Release audio focus since the dialog prevents keyboards from controlling memprime.
                mActivity!!.maybeChangeAudioFocus(false)
            }
            TtsSpeaker.speak("Great job! Deck done.")
            val deckChooser = DeckChooseModeSetter.getInstance()
            val nextDeckWithItems = deckChooser.nextDeckWithItems
            if (nextDeckWithItems != null) {
                deckChooser.loadDeck(nextDeckWithItems)
                this.switchMode(mActivity!!)
                TtsSpeaker.speak("Loading " + nextDeckWithItems.name)
            } else {
                TtsSpeaker.speak("All decks done..")
            }
        }
        commonLayoutSetup()
        memoryDroid!!.findViewById<View>(R.id.rerecord)
            .setOnTouchListener(
                RecordOnClickListener(
                    currentNote, context, false,
                    lastOrNull
                )
            )
    }

    private val lastOrNull: Note?
        get() = if (!lastNoteList.isEmpty()) lastNoteList.last else null

    private fun setupAnswerMode(context: Activity) {
        questionMode = false
        memoryDroid!!.setContentView(R.layout.learnquestion)
        applyDim(mIsDimmed)
        val currentNote = currentNote
        val lifeCycleOwner = mActivity
        if (currentNote != null && lifeCycleOwner != null) {
            MoveManager.replaceMoveJobWith(lifeCycleOwner.lifecycleScope.launch(Dispatchers.Main) {
                AudioPlayer.instance.playFile(currentNote.answer, shouldRepeat = true)
                delay(30_000)
                if (isActive) {
                    TtsSpeaker.speak("Auto-proceed from answer.")
                    proceed()
                }
            })
        }
        commonLayoutSetup()
        memoryDroid!!.findViewById<View>(R.id.rerecord)
            .setOnTouchListener(
                RecordOnClickListener(
                    currentNote, context, true,
                    lastOrNull
                )
            )

    }


    private fun updateVal() {
        currentNote = currentDeckReviewQueue!!.peekQueue()
        if (currentNote != null) {
            repCounter++
            if (repCounter % 10 == 9) {
                markAllStale(mActivity!!)
            }
        }
    }

    private fun updateScoreAndMoveToNext(context: Activity, newGrade: Int) {
        if (currentNote != null) {
            applyGradeStatic(context, newGrade, currentNote!!)
            ToastSingleton.getInstance()
                .msg("Easiness: " + currentNote!!.easiness + " Interval " + currentNote!!.interval)
            setupQuestionMode(context)
        }
    }

    private fun applyGradeStatic(
        context: Activity?, newGrade: Int,
        currentNote: Note
    ) {
        val noteEditor = DbNoteEditor.instance
        if (lastNoteRep != null && lastOrNull != null) {
            val repEditor = DbRepEditor.getInstance()
            repEditor.insert(lastNoteRep)
        }
        // It seems unwise to create a clone, which not just update the current one.
        lastNoteList.addLast(currentNote.clone())
        // Create the rep info before updating the note with the new internal.
        lastNoteRep = AbstractRep(
            currentNote.id,
            currentNote.interval, newGrade, System.currentTimeMillis()
        )
        currentNote.process_answer(newGrade)
        noteEditor!!.update(context!!, currentNote)

        // If you scored too low review it again, at the end.
        if (currentNote.is_due_for_acquisition_rep) {
            currentDeckReviewQueue!!.updateNote(currentNote, false)
            missCounter++
        } else {
            currentDeckReviewQueue!!.removeNote(currentNote.id)
        }
    }

    override fun deleteNote() {
        val noteEditor = DbNoteEditor.instance
        val note = currentNote
        if (note != null) {
            noteEditor!!.deleteCurrent(mActivity!!, note)
            currentDeckReviewQueue!!.removeNote(note.id)
        }
        setupQuestionMode(mActivity!!)
    }

    private fun replayA() {
        ScreenDimmer.getInstance().keepScreenOn(mActivity)
        if (currentNote != null) {
            AudioPlayer.instance.playFile(currentNote!!.answer, null, true)
        }
    }

    private fun replay() {
        ScreenDimmer.getInstance().keepScreenOn(mActivity)
        if (currentNote != null) {
            AudioPlayer.instance.playFile(currentNote!!.question, null, true)
        }
    }

    private fun proceed(context: Activity) {
        if (currentNote != null) {
            setupAnswerMode(context)
        }
    }

    private fun undoLastQuestion(context: Activity) {
        if (!lastNoteList.isEmpty()) {
            val currentNote = lastNoteList.removeLast()
            this.currentNote = currentNote
            repCounter--
            val noteEditor = DbNoteEditor.instance
            noteEditor!!.update(context, currentNote)
            // In case the grade was bad take it out of revision queue.
            currentDeckReviewQueue!!.removeNote(currentNote.id)
            currentDeckReviewQueue!!.addToFront(currentNote)
            setupAnswerMode(context)
        } else {
            TtsSpeaker.speak("Nothing to undo")
        }
    }

    override fun resetActivity() {
        switchMode(mActivity!!)
    }

    private fun undoThisQuestion(context: Activity) {
        setupQuestionMode(context, false)
    }

    override fun adjustScreenLock() {
        ScreenDimmer.getInstance().keepScreenOn(mActivity)
        hideSystemUi()
        mIsDimmed = false
    }

    private var mIsDimmed = false
    override fun toggleDim() {
        mIsDimmed = !mIsDimmed
        applyDim(mIsDimmed)
        showSystemUi()
    }

    override fun mark() {
        if (currentNote != null) {
            currentNote!!.isMarked = true
            val noteEditor = DbNoteEditor.instance
            noteEditor!!.update(mActivity!!, currentNote!!)
            ToastSingleton.getInstance().msg("Marked note")
        }
    }

    private fun applyDim(isDimmed: Boolean) {
        // Note making the root invisible makes the background grey.
        val layout = memoryDroid!!.findViewById<ViewGroup>(R.id.learn_layout_root)
        for (i in 0 until layout.childCount) {
            applyDimToView(isDimmed, layout.getChildAt(i))
        }
        applyDimToView(isDimmed, memoryDroid!!.findViewById(R.id.reviewNumber))

        // Gestures listeners don't work if dimmed.
        applyDimToView(false, memoryDroid!!.findViewById(R.id.gestures))
    }

    private fun applyDimToView(isDimmed: Boolean, view: View) {
        if (!isDimmed) {
            view.visibility = View.VISIBLE
        } else {
            view.visibility = View.INVISIBLE
        }
    }
}